### PR TITLE
[#1180] `VaDataTable` grid mode

### PR DESCRIPTION
**Pull request by aluarius**
Close: #1180 

## Description
- [x] added grid view mode;
- [x] docs updates;
- [x] refactoring;
- [x] minor bugfixing (have faced several old bugs). 

image

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

### CODE REVIEW

File: packages/ui/src/components/va-data-table/hooks/useCommonProps.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { PropType } from 'vue'
+
+import { DataTableItem } from '../types'
+
+export const currentPage = { currentPage: { type: Number as PropType<number | undefined> } }
+
+export const items = { items: { type: Array as PropType<DataTableItem[]>, default: () => [] as DataTableItem[] } }
```
Comment (rustem-nasyrov) [inline packages/ui/src/components/va-data-table/hooks/useCommonProps.ts:7]: This kind of naming might mislead. Would be better to name props like others in the `VaDataTable` component. For example `useDataTableItems`.

File: packages/ui/src/components/va-data-table/hooks/useTableScroll.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ export const useTableScrollProps = {
 
 export const useTableScrollEmits = ['scroll:top', 'scroll:bottom']
 
-export default function useTableScroll (
+export function useTableScroll (
   props: ExtractPropTypes<typeof useTableScrollProps>,
   emit: (event: 'scroll:bottom' | 'scroll:top', ...args: any[]) => void,
 ) {
```
Comment (rustem-nasyrov) [inline packages/ui/src/components/va-data-table/hooks/useTableScroll.ts:15]:
```suggestion
export const useTableScroll = (
  props: ExtractPropTypes<typeof useTableScrollProps>,
  emit: (event: 'scroll:bottom' | 'scroll:top', ...args: any[]) => void,
) => {
```

File: packages/ui/src/components/va-data-table/hooks/useStylable.ts [hunk 3]
```diff
@@ -14,19 +15,19 @@ const prefix = '--va-data-table'
...
+  stickyHeader: { type: Boolean, default: false },
+  stickyFooter: { type: Boolean, default: false },
+  height: { type: [String, Number] as PropType<string | number | undefined> },
 }
 
 const getClass = (classes: DataTableColumnClass) => isFunction(classes) ? classes() : classes
 const getStyle = (styles: DataTableColumnStyle) => isFunction(styles) ? styles() : styles
 
-export default function useStylable (props: useStylableProps) {
+export function useStylable (props: ExtractPropTypes<typeof useStylableProps>) {
```
Comment (rustem-nasyrov) [inline packages/ui/src/components/va-data-table/hooks/useStylable.ts:30]:
```suggestion
export const useStylable = (props: ExtractPropTypes<typeof useStylableProps>) => {
```

File: packages/ui/src/components/va-data-table/hooks/usePaginatedRows.ts [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
+  ...currentPage,
+  perPage: { type: Number as PropType<number | undefined> },
 }
 
-export default function usePaginatedRows (
+export function usePaginatedRows (
   sortedRows: Ref<DataTableRow[]>,
-  props: useSelectableProps,
+  props: ExtractPropTypes<typeof usePaginatedRowsProps>,
 ) {
```
Comment (rustem-nasyrov) [inline packages/ui/src/components/va-data-table/hooks/usePaginatedRows.ts:15]:
```suggestion
export const usePaginatedRows = (
  sortedRows: Ref<DataTableRow[]>,
  props: ExtractPropTypes<typeof usePaginatedRowsProps>,
) => {

File: packages/ui/src/components/va-data-table/hooks/useAnimationName.ts [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
+  ...currentPage,
+  animated: { type: Boolean, default: true },
 }
 
-export default function useAnimationName (
-  props: useAnimationNameProps,
+export function useAnimationName (
+  props: ExtractPropTypes<typeof useAnimationNameProps>,
   rows: Ref<DataTableRow[]>,
 ) {
```
Comment (rustem-nasyrov) [inline packages/ui/src/components/va-data-table/hooks/useAnimationName.ts:17]:
```suggestion
export const useAnimationName = (
  props: ExtractPropTypes<typeof useAnimationNameProps>,
  rows: Ref<DataTableRow[]>,
) => {
```

**Review by rustem-nasyrov — CHANGES_REQUESTED** [review]
Great, but still things to do. Also, I'll check functionality a bit later.

File: packages/ui/src/components/va-data-table/components/VaDataTableThRow.vue [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        </div>
+    </th>
+  </tr>
+</template>
+
+<script lang="ts">
+import { defineComponent, PropType, computed } from 'vue'
+
+import { VaIcon } from '../../va-icon'
+import { VaCheckbox } from '../../va-checkbox'
```
[the remark this answers is not in the document]
Comment (aluarius) [reply packages/ui/src/components/va-data-table/components/VaDataTableThRow.vue:69]: It ruins precommit check.

**Review by m0ksem — CHANGES_REQUESTED** [review]
image

As I understand gird can be helpful in mobile, but....

Add breakpoints here

**Comment by m0ksem** [comment]
```css
--va-data-table-grid-count: unset

@media for mobile {
  --va-actual-data-table-grid-count: var(--va-data-table-grid-count, 2)
}

@media for desktop{
  --va-actual-data-table-grid-count: var(--va-data-table-grid-count, 4)
}
```

**Review by rustem-nasyrov — APPROVED** [review]

**Comment by m0ksem** [comment]
I don't think it closes https://github.com/epicmaxco/vuestic-ui/issues/2338

**Pull request merged**